File: src/main/java/com/lambda/schematic/Schematic.java
package com.lambda.schematic;

import net.minecraft.block.state.IBlockState;
import net.minecraft.util.math.BlockPos;

public interface Schematic {
    default IBlockState desiredState(int x, int y, int z) {
        return desiredState(new BlockPos(x, y, z));
    }

    IBlockState desiredState(BlockPos pos);

    default boolean inSchematic(BlockPos pos) {
        return inSchematic(pos.getX(), pos.getY(), pos.getZ());
    }

    default boolean inSchematic(int x, int y, int z) {
        return x >= 0 && x < widthX() && y >= 0 && y < heightY() && z >= 0 && z < lengthZ();
    }

    int widthX();

    int heightY();

    int lengthZ();
}

File: src/main/kotlin/com/lambda/client/util/schematic/Schematic.kt
package com.lambda.client.util.schematic

import net.minecraft.block.state.IBlockState
import net.minecraft.util.math.BlockPos

interface Schematic {
    fun desiredState(x: Int, y: Int, z: Int): IBlockState {
        return desiredState(BlockPos(x, y, z))
    }

    fun desiredState(pos: BlockPos): IBlockState

    fun inSchematic(pos: BlockPos): Boolean
    fun inSchematic(x: Int, y: Int, z: Int): Boolean

    fun widthX(): Int
    fun heightY(): Int
    fun lengthZ(): Int
    fun getOrigin(): BlockPos
}